subsys/net/lib/lwm2m/ucifi_battery.c

```diff
@@ -0,0 +1,150 @@
+	OBJ_FIELD_DATA(UCIFI_BATTERY_NUM_CYCLES_RID, R_OPT, U32),
+	OBJ_FIELD_DATA(UCIFI_SUPPLY_LOSS_RID, R_OPT, BOOL),
+	OBJ_FIELD_DATA(UCIFI_SUPPLY_LOSS_COUNTER_RID, R_OPT, U32),
+	OBJ_FIELD_EXECUTE_OPT(UCIFI_SUPPLY_LOSS_COUNTER_RESET_RID),
+	OBJ_FIELD_DATA(UCIFI_SUPPLY_LOSS_REASON_RID, R_OPT, STRING),
+};
+
+static struct lwm2m_engine_obj_inst inst[MAX_INSTANCE_COUNT];
+static struct lwm2m_engine_res res[MAX_INSTANCE_COUNT][BATTERY_MAX_ID];
+static struct lwm2m_engine_res_inst res_inst[MAX_INSTANCE_COUNT][RESOURCE_INSTANCE_COUNT];
+
+static void clear_supply_loss_counter(uint16_t obj_inst_id, int index)
+{
+	supply_loss_counter[index] = 0;
+	NOTIFY_OBSERVER(UCIFI_OBJECT_BATTERY_ID, obj_inst_id, UCIFI_SUPPLY_LOSS_COUNTER_RID);
+}
+
+static int supply_loss_counter_reset_cb(uint16_t obj_inst_id, uint8_t *args, uint16_t args_len)
+{
+	int i;
+
+	LOG_DBG("RESET supply loss counter %d", obj_inst_id);
+	for (i = 0; i < MAX_INSTANCE_COUNT; i++) {
+		if (inst[i].obj && inst[i].obj_inst_id == obj_inst_id) {
+			clear_supply_loss_counter(obj_inst_id, i);
+			return 0;
+		}
+	}
+
+	return -ENOENT;
+}
+
+static struct lwm2m_engine_obj_inst *battery_create(uint16_t obj_inst_id)
+{
+	int index, i = 0, j = 0;
+
+	/* Check that there is no other instance with this ID */
+	for (index = 0; index < MAX_INSTANCE_COUNT; index++) {
+		if (inst[index].obj && inst[index].obj_inst_id == obj_inst_id) {
+			LOG_ERR("Can not create instance - "
+				"already existing: %u",
+				obj_inst_id);
+			return NULL;
+		}
+	}
+
+	for (index = 0; index < MAX_INSTANCE_COUNT; index++) {
+		if (!inst[index].obj) {
+			break;
+		}
+	}
+
+	if (index >= MAX_INSTANCE_COUNT) {
+		LOG_ERR("Can not create instance - no more room: %u", obj_inst_id);
+		return NULL;
+	}
+
+	/* Set default values */
+	battery_level[index] = 0;
+	supply_loss_counter[index] = 0;
+
+	(void)memset(res[index], 0, sizeof(res[index][0]) * ARRAY_SIZE(res[index]));
+	init_res_instance(res_inst[index], ARRAY_SIZE(res_inst[index]));
+
+	/* initialize instance resource data */
+	INIT_OBJ_RES_DATA(UCIFI_BATTERY_LEVEL_RID, res[index], i, res_inst[index], j,
+			  &battery_level[index], sizeof(*battery_level));
+	INIT_OBJ_RES_OPTDATA(UCIFI_BATTERY_CAPACITY_RID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(UCIFI_BATTERY_VOLTAGE_RID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(UCIFI_BATTERY_TYPE_RID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(UCIFI_BATTERY_LOW_THESHOLD_RID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(UCIFI_BATTERY_LEVEL_TOO_LOW_RID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(UCIFI_BATTERY_SHUTDOWN_RID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(UCIFI_BATTERY_NUM_CYCLES_RID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(UCIFI_SUPPLY_LOSS_RID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(UCIFI_SUPPLY_LOSS_COUNTER_RID, res[index], i, res_inst[index], j);
```

Comment:
I was wondering about this. I didn't because I didn't want to include the optional resource by default. But because of the execute, probably need to do it that way.